bevy_reflect: get_represented_kind_info APIs for reflected kinds

torsteingrindvik opened the pull request:
# Objective

Fixes #14378

alice-i-cecile commented on crates/bevy_reflect/src/struct_trait.rs:
```diff
@@ -73,6 +73,10 @@ pub trait Struct: Reflect {
+
+    /// Returns the structs [`StructInfo`].
+    /// Will return `None` if [`TypeInfo`] is not available.
+    fn get_represented_kind_info(&self) -> Option<&StructInfo>;
```
We could make this a method on `Reflect` and then use an associated type to control the return value, right? Would that be better?

alice-i-cecile replied:
Hmm, actually, that's a bad idea. Because then we can't do a `Box<dyn Reflect>` without specifying the associated type.

MrGVSV commented on crates/bevy_reflect/src/lib.rs:
```diff
@@ -1751,6 +1751,18 @@ mod tests {
+        #[derive(Reflect)]
+        struct SomeStruct;
+
+        let ReflectRef::Struct(dyn_struct) = SomeStruct.reflect_ref() else {
+            unreachable!()
+        };
```
Nit: you could probably simplify this
```suggestion
        let dyn_struct: &dyn Struct = &SomeStruct;
```

MrGVSV commented on crates/bevy_reflect/src/struct_trait.rs:
```diff
@@ -73,6 +73,10 @@ pub trait Struct: Reflect {
+
+    /// Returns the structs [`StructInfo`].
+    /// Will return `None` if [`TypeInfo`] is not available.
+    fn get_represented_kind_info(&self) -> Option<&StructInfo>;
```
Could we make this return a `'static` reference?

torsteingrindvik replied:
Ah I didn't realize that would be available. Changed

MrGVSV commented on crates/bevy_reflect/src/struct_trait.rs:
```diff
@@ -73,6 +73,10 @@ pub trait Struct: Reflect {
+
+    /// Returns the structs [`StructInfo`].
+    /// Will return `None` if [`TypeInfo`] is not available.
+    fn get_represented_kind_info(&self) -> Option<&StructInfo>;
```
Hm, now I also wonder if we should maybe give these each kind-specific names? That might make it easier to find and reason about. 🤔

For example, this could be `get_represented_struct_info`.

Thoughts?

alice-i-cecile replied:
Yeah, I think that's a bit less error prone.

torsteingrindvik replied:
I'll defer to you but: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

Any reason to not go for the short and simple `.struct_info()`?

MrGVSV replied:
> I'll defer to you but: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter
> 
> 
> 
> Any reason to not go for the short and simple `.struct_info()`?

I do prefer the shorter version, but for now I think we should stick with the longer version so that things are consistent.

In a followup, maybe we rename all the `get_represented_***` methods to their shorter versions?

Originally I had it to better differentiate with the method on `Typed` and to make it clear that this is the _represented_ info. But with proper documentation, I think can simplify (in a breaking-change followup PR).

torsteingrindvik replied:
Understandable :+1: 

Pushed the rest of the impls (except `ValueInfo`, not sure if that fits into this) using the long form names for now.

I don't mind doing the renaming chore in a follow up later if it's desired to do it. Imo. the long names are a bit noisy.

torsteingrindvik replied:
> directly implemented inherently on `dyn Struct`

Not sure I'm following- could you elaborate?

MrGVSV commented on crates/bevy_reflect/src/struct_trait.rs:
```diff
@@ -73,6 +73,10 @@ pub trait Struct: Reflect {
     /// Clones the struct into a [`DynamicStruct`].
     fn clone_dynamic(&self) -> DynamicStruct;
+
+    /// Returns the structs [`StructInfo`].
+    /// Will return `None` if [`TypeInfo`] is not available.
```
Nit:
```suggestion
    /// Returns the struct's [`StructInfo`].
    ///
    /// Will return `None` if [`TypeInfo`] is not available.
```

MrGVSV commented on crates/bevy_reflect/derive/src/impls/structs.rs:
```diff
@@ -129,6 +129,11 @@ pub(crate) fn impl_struct(reflect_struct: &ReflectStruct) -> proc_macro2::TokenS
                 dynamic
             }
+
+            fn get_represented_kind_info(&self) -> #FQOption<&#bevy_reflect_path::StructInfo> {
+                let type_info = #bevy_reflect_path::Reflect::get_represented_type_info(self)?;
+                type_info.as_struct().ok()
+            }
```
Alternatively we can probably have the trait supply a default implementation. I’m not sure which is better or if the compiler prefers one over the other for optimization purposes.

But putting it in a default method at least reduces the amount of generated code.

torsteingrindvik replied:
Went for the default impl

torsteingrindvik commented:
@alice-i-cecile @MrGVSV  any more concerns? I believe I addressed all review comments, the only open discussion was about naming- and my impression of that thread is that it's non-blocking.

pablo-lua commented on crates/bevy_reflect/src/enums/enum_trait.rs:
```diff
@@ -130,6 +130,13 @@ pub trait Enum: Reflect {
+
+    /// Will return `None` if [`TypeInfo`] is not available.
+    ///
+    /// [`TypeInfo`]: crate::TypeInfo
```
Docs isn't consistent with the other methods, is there a reason that here we have this second line but in other places not?

torsteingrindvik replied:
In doc comments I've seen in Bevy the convention is to prefer the shorter syntax without path qualifiers.
But doing so requires the type to be in scope.
This file did not have `TypeInfo` in scope, the others did.
Rust allows bringing types into scope via the syntax you see on the last line.

pablo-lua reviewed (approved):
I have just one question, but isn't really blocking.
